fvtest/compilertriltest/ShiftAndRotateTest.cpp

```diff
@@ -445,3 +1011,431 @@ INSTANTIATE_TEST_CASE_P(ShiftAndRotateTest, UInt16ShiftAndRotate, ::testing::Com
+        "  (block"
+        "    (lreturn"
+        "      (%s"
+        "        (land"
+        "          (lload parm=0)"
+        "          (lconst %" PRIu64 "))"
+        "        (iconst %d)))))",
+        param.opcode.c_str(),
+        param.mask,
+        param.shift);
+    auto trees = parseString(inputTrees);
+
+    ASSERT_NOTNULL(trees);
+
+    Tril::DefaultCompiler compiler(trees);
+
+    ASSERT_EQ(0, compiler.compile()) << "Compilation failed unexpectedly\n" << "Input trees: " << inputTrees;
+
+    auto entry_point = compiler.getEntryPoint<int64_t (*)(int64_t)>();
+    ASSERT_EQ(param.oracle(param.value, param.mask, param.shift), entry_point(param.value));
+}
+
+INSTANTIATE_TEST_CASE_P(ShiftAndRotateTest, Int64MaskThenShift, ::testing::Combine(
+    ::testing::ValuesIn(static_cast< std::vector<std::tuple<int64_t, int64_t, int32_t>> (*) (void)>(test_mask_then_shift_values)()),
+    ::testing::Values(std::make_tuple<const char*, int64_t (*) (int64_t, int64_t, int32_t)>("lshr", static_cast<int64_t (*) (int64_t, int64_t, int32_t)>(mask_then_shift_right))
+    )));
+
+class UInt32MaskThenShift : public MaskThenShiftArithmetic<uint32_t> {};
+
+TEST_P(UInt32MaskThenShift, UsingLoadParam) {
+    auto param = to_mask_then_shift_struct(GetParam());
+
+    char inputTrees[300] = {0};
+    std::snprintf(inputTrees, sizeof(inputTrees),
+        "(method return=Int32 args=[Int32]"
+        "  (block"
+        "    (ireturn"
+        "      (%s"
+        "        (iand"
+        "          (iload parm=0)"
+        "          (iconst %u))"
+        "        (iconst %d)))))",
+        param.opcode.c_str(),
+        param.mask,
+        param.shift);
+    auto trees = parseString(inputTrees);
+
+    ASSERT_NOTNULL(trees);
+
+    Tril::DefaultCompiler compiler(trees);
+
+    ASSERT_EQ(0, compiler.compile()) << "Compilation failed unexpectedly\n" << "Input trees: " << inputTrees;
+
+    auto entry_point = compiler.getEntryPoint<uint32_t (*)(uint32_t)>();
+    ASSERT_EQ(param.oracle(param.value, param.mask, param.shift), entry_point(param.value));
+}
+
+INSTANTIATE_TEST_CASE_P(ShiftAndRotateTest, UInt32MaskThenShift, ::testing::Combine(
+    ::testing::ValuesIn(static_cast< std::vector<std::tuple<uint32_t, uint32_t, int32_t>> (*) (void)>(test_mask_then_shift_values)()),
+    ::testing::Values(std::make_tuple<const char*, uint32_t (*) (uint32_t, uint32_t, int32_t)>("iushr", static_cast<uint32_t (*) (uint32_t, uint32_t, int32_t)>(mask_then_shift_right))
+    )));
+
+class Int32MaskThenShift : public MaskThenShiftArithmetic<int32_t> {};
+
+TEST_P(Int32MaskThenShift, UsingLoadParam) {
+    auto param = to_mask_then_shift_struct(GetParam());
+
+    char inputTrees[300] = {0};
+    std::snprintf(inputTrees, sizeof(inputTrees),
+        "(method return=Int32 args=[Int32]"
+        "  (block"
+        "    (ireturn"
+        "      (%s"
+        "        (iand"
+        "          (iload parm=0)"
+        "          (iconst %u))"
```

Comment:
It seems a bit odd to be using `%u` here because this test deals with `int32_t`, but it should still technically work.

Reply:
Looks like I accidentally switched this when I was reformatting the trees and didn't notice. Fixed in 7836d6374b4279253ad4821df545ef411d8d46cb.